packages/trimerge-sync/src/TrimergeClient.ts

```diff
@@ -272,56 +279,61 @@ export class TrimergeClient<
         const {
           doc,
           editMetadata,
-          lazy = false,
+          temp = true,
         } = this.differ.merge(base, left, right);
         return this.addNewCommit(
           doc,
           editMetadata,
-          lazy,
+          temp,
           left,
           rightRef,
           baseRef,
         );
       },
     );
+    if (this.allHeadRefs.size > 1) {
+      throw new Error('more than one head after merging');
+    }
     // TODO: update Presence(s) based on this merge
     // TODO: can we clear out commits we don't need anymore?
   }
 
-  private syncPromise: Promise<boolean> | undefined;
-
   private emitClientListChange() {
     this.clientList = Array.from(this.clientMap.values());
     this.clientListSubs.emitChange();
   }
-
-  private get needsSync(): boolean {
-    return this.unsyncedCommits.length > 0 || this.newPresence !== undefined;
-  }
   private sync(): void {
-    if (!this.syncPromise && this.needsSync) {
-      this.syncPromise = this.doSync();
+    const commits = this.unsyncedCommits;
+    if (commits.length > 0 || this.newPresence !== undefined) {
+      this.unsyncedCommits = [];
+      this.updateSyncState({ localSave: 'saving' });
+      this.store.update(commits, this.newPresence);
+      this.newPresence = undefined;
     }
+    this.updateSyncState({ localSave: 'ready' });
   }
 
-  private updateSyncState(update: Partial<SyncStatus>) {
+  private updateSyncState(update: Partial<SyncStatus>): void {
     this.syncState = { ...this.syncState, ...update };
     this.syncStateSubs.emitChange();
   }
-
-  private async doSync() {
-    while (this.needsSync) {
-      this.updateSyncState({ localSave: 'pending' });
-      await timeout(this.bufferMs);
-      const commits = this.unsyncedCommits;
-      this.unsyncedCommits = [];
-      this.updateSyncState({ localSave: 'saving' });
-      this.store.update(commits, this.newPresence);
-      this.newPresence = undefined;
+  private addHead(
```

Comment:
nit: I feel like it makes sense for the set that you're adding should come first.